flocker/node/agents/ebs.py

```diff
@@ -900,6 +900,91 @@ def _get_blockdevices():
     return FilePath(b"/sys/block").children()
 
 
+def _next_device():
+    """
+    Get the next available EBS device name for this EC2 instance.
+
+    :return Optional[unicode] file_name: available device name for
+        attaching EBS volume, or None if suitable EBS device names on this
+        EC2 instance are currently occupied
+    """
+
+    return _select_free_device(_find_allocated_devices())
+
+
+def _find_allocated_devices():
+    """
+    Enumerate the allocated device names on this host.
+
+    :return Sequence[bytes]: List of allocated device basenames
+        (e.g. ``[b'sda']``).
+    """
+    command = [
+        b"/bin/lsblk",
+        b"--nodeps",           # Only print top-level devices
+        b"--noheadings",       # Do not display KNAME header
+        b"--output", b"KNAME"  # Only display kernel device name
+    ]
+    command_result = check_output(command)
+    existing = [
+        dev for dev in command_result.split("\n")
+        if dev.startswith(b"xvd") or dev.startswith(b'sd')
+    ]
+    return existing
+
+
+def _select_free_device(existing):
+    """
+    Given a list of allocated devices, return an available device name.
+
+    According to
+    http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html
+    all AWS Linux instances have ``/dev/sd[a-z]`` available.  However:
+
+    - ``sda`` is reserved for the root device (technically only ``sda1``);
+    - Amazon "strongly recommend that you don't" use instance store names
+      (usually ``/dev/sd[b-e]``) "because the behavior can be unpredictable";
+    - some "custom kernels might have restrictions that limit use to
+      ``/dev/sd[f-p]``".
+
+    ``sd[f-p]`` only allows 11 devices, so to increase this, ignore the
+    least stringent statement above, and allow ``sd[f-z]`` (21 devices).
+
+    To reduce the risk of failing on custom AMIs, select from ``[f-p]`` first.
+
+    Any further increase will need to start mining the ``hd[a-z][1-15]``
+    and ``xvd[b-c][a-z]`` namespaces, but which to use depends on whether
+    the AMI uses paravirtualization or HVM.
+
+    :param Sequence[bytes]: List of allocated device basenames
+        (e.g. ``[b'sda']``).
+    :return Optional[unicode] file_name: available device name for
+        attaching EBS volume, or None if suitable EBS device names on this
+        EC2 instance are currently occupied
+    """
+
+    local_devices = pset(existing)
```

Comment:
:question: The logic below would be simplified by just using `local_devices` and doing `if not possible_devices.intersection(local_devices)`.